phase4-entsog-client/src/main/java/com/helger/phase4/entsog/Phase4ENTSOGSender.java

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright (C) 2020-2021 Philip Helger (www.helger.com)
+ * philip[at]helger[dot]com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.helger.phase4.entsog;
+
+import java.nio.charset.Charset;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.OverridingMethodsMustInvokeSuper;
+import javax.annotation.concurrent.Immutable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.helger.phase4.attachment.Phase4OutgoingAttachment;
+import com.helger.phase4.attachment.WSS4JAttachment;
+import com.helger.phase4.client.AS4ClientUserMessage;
+import com.helger.phase4.crypto.ECryptoKeyIdentifierType;
+import com.helger.phase4.sender.AS4BidirectionalClientHelper;
+import com.helger.phase4.sender.AbstractAS4UserMessageBuilder;
+import com.helger.phase4.util.AS4ResourceHelper;
+import com.helger.phase4.util.Phase4Exception;
+
+/**
+ * This class contains all the specifics to send AS4 messages with the ENTSOG
+ * profile. See <code>sendAS4Message</code> as the main method to trigger the
+ * sending, with all potential customization.
+ *
+ * @author Pavel Rotek
+ * @since 0.13.3
+ */
+@Immutable
+public final class Phase4ENTSOGSender {
+	private static final Logger LOGGER = LoggerFactory.getLogger(Phase4ENTSOGSender.class);
+
+	private Phase4ENTSOGSender() {
+	}
+
+	/**
+	 * @return Create a new Builder for AS4 messages if the payload is present.
+	 *         Never <code>null</code>.
+	 */
+	@Nonnull
+	public static ENTSOGUserMessageBuilder builder() {
+		return new ENTSOGUserMessageBuilder();
+	}
+
+	/**
+	 * Abstract ENTSOG UserMessage builder class with sanity methods
+	 *
+	 * @author Philip Helger
+	 * @param <IMPLTYPE> The implementation type
+	 */
+	public abstract static class AbstractENTSOGUserMessageBuilder<IMPLTYPE extends AbstractENTSOGUserMessageBuilder<IMPLTYPE>>
+			extends AbstractAS4UserMessageBuilder<IMPLTYPE> {
+
+		private ECryptoKeyIdentifierType m_signingKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
```

Comment:
I guess I would prefer if you make a constant like
```
public static final ECryptoKeyIdentifierType  DEFAULT_KEY_IDENTIFIER_TYPE = ECryptoKeyIdentifierType.ISSUER_SERIAL;
```
and than call the setters in the constructor. That would be kind of consistent with the other builders.

Reply:
depends on if we move it higher up the hierarchy or how we solve cryptoparams customization globally